The notebook to explore different processing (scaling) order to the final result, I will discuss 
the following options:

For Deep Learning (Regression Problem only):

    - Scaling whole datasets for Deep Learning model
    - Scaling separate training and testing datasets for Deep Learning model

For Machine Learning (Regression Problem only):
    
    - Scaling whole datasets for Machine Learning model
    - Scaling separate training and testing datasets for Machine Learning model

In [1]:
import numpy as np
import pandas as pd

from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
# diaplay all the columns
pd.pandas.set_option("display.max_columns", None)

data = pd.read_csv("train.csv")

data.drop("Id", axis=1, inplace=True)


In [2]:
# get the categorical and numerical featuers
car_feas = [fea for fea in data.columns if data[fea].dtype == 'O']
## add MSSubClass to cate features
car_feas = car_feas + ['MSSubClass']
num_feas = [fea for fea in data.columns if fea not in car_feas and fea not in ["SalePrice"]]
# get and test on numeric data only
num_data = data[num_feas]

In [3]:
num_data

LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  YearRemodAdd  \
0            65.0     8450            7            5       2003          2003   
1            80.0     9600            6            8       1976          1976   
2            68.0    11250            7            5       2001          2002   
3            60.0     9550            7            5       1915          1970   
4            84.0    14260            8            5       2000          2000   
...           ...      ...          ...          ...        ...           ...   
1455         62.0     7917            6            5       1999          2000   
1456         85.0    13175            6            6       1978          1988   
1457         66.0     9042            7            9       1941          2006   
1458         68.0     9717            5            6       1950          1996   
1459         75.0     9937            5            6       1965          1965   

      MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  1stFlrSF  \
0          196.0         706           0        150          856       856   
1            0.0         978           0        284         1262      1262   
2          162.0         486           0        434          920       920   
3            0.0         216           0        540          756       961   
4          350.0         655           0        490         1145      1145   
...          ...         ...         ...        ...          ...       ...   
1455         0.0           0           0        953          953       953   
1456       119.0         790         163        589         1542      2073   
1457         0.0         275           0        877         1152      1188   
1458         0.0          49        1029          0         1078      1078   
1459         0.0         830         290        136         1256      1256   

      2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  \
0          854             0       1710             1             0         2   
1            0             0       1262             0             1         2   
2          866             0       1786             1             0         2   
3          756             0       1717             1             0         1   
4         1053             0       2198             1             0         2   
...        ...           ...        ...           ...           ...       ...   
1455       694             0       1647             0             0         2   
1456         0             0       2073             1             0         2   
1457      1152             0       2340             0             0         2   
1458         0             0       1078             1             0         1   
1459         0             0       1256             1             0         1   

      HalfBath  BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd  Fireplaces  \
0            1             3             1             8           0   
1            0             3             1             6           1   
2            1             3             1             6           1   
3            0             3             1             7           1   
4            1             4             1             9           1   
...        ...           ...           ...           ...         ...   
1455         1             3             1             7           1   
1456         0             3             1             7           2   
1457         0             4             1             9           2   
1458         0             2             1             5           0   
1459         1             3             1             6           0   

      GarageYrBlt  GarageCars  GarageArea  WoodDeckSF  OpenPorchSF  \
0          2003.0           2         548           0           61   
1          1976.0           2         460         298            0   
2          2001.0           2         608       

1. Scaling First for ML Model

In [4]:
# processing data
for fea in num_feas:
    mean_val = data[fea].mean()
    # add binary missing indicator
    num_data[fea+'_na'] = np.where(num_data[fea].isnull(), 1, 0)
    # replace missing values by the mean
    num_data[fea].fillna(mean_val, inplace=True)

<ipython-input-4-6c23c6ee0c4c>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  num_data[fea+'_na'] = np.where(num_data[fea].isnull(), 1, 0)
C:\Users\ren\AppData\Local\Programs\Python\Python38\lib\site-packages\pandas\core\series.py:4517: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().fillna(


In [5]:
num_data

LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  YearRemodAdd  \
0            65.0     8450            7            5       2003          2003   
1            80.0     9600            6            8       1976          1976   
2            68.0    11250            7            5       2001          2002   
3            60.0     9550            7            5       1915          1970   
4            84.0    14260            8            5       2000          2000   
...           ...      ...          ...          ...        ...           ...   
1455         62.0     7917            6            5       1999          2000   
1456         85.0    13175            6            6       1978          1988   
1457         66.0     9042            7            9       1941          2006   
1458         68.0     9717            5            6       1950          1996   
1459         75.0     9937            5            6       1965          1965   

      MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  1stFlrSF  \
0          196.0         706           0        150          856       856   
1            0.0         978           0        284         1262      1262   
2          162.0         486           0        434          920       920   
3            0.0         216           0        540          756       961   
4          350.0         655           0        490         1145      1145   
...          ...         ...         ...        ...          ...       ...   
1455         0.0           0           0        953          953       953   
1456       119.0         790         163        589         1542      2073   
1457         0.0         275           0        877         1152      1188   
1458         0.0          49        1029          0         1078      1078   
1459         0.0         830         290        136         1256      1256   

      2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  \
0          854             0       1710             1             0         2   
1            0             0       1262             0             1         2   
2          866             0       1786             1             0         2   
3          756             0       1717             1             0         1   
4         1053             0       2198             1             0         2   
...        ...           ...        ...           ...           ...       ...   
1455       694             0       1647             0             0         2   
1456         0             0       2073             1             0         2   
1457      1152             0       2340             0             0         2   
1458         0             0       1078             1             0         1   
1459         0             0       1256             1             0         1   

      HalfBath  BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd  Fireplaces  \
0            1             3             1             8           0   
1            0             3             1             6           1   
2            1             3             1             6           1   
3            0             3             1             7           1   
4            1             4             1             9           1   
...        ...           ...           ...           ...         ...   
1455         1             3             1             7           1   
1456         0             3             1             7           2   
1457         0             4             1             9           2   
1458         0             2             1             5           0   
1459         1             3             1             6           0   

      GarageYrBlt  GarageCars  GarageArea  WoodDeckSF  OpenPorchSF  \
0          2003.0           2         548           0           61   
1          1976.0           2         460         298            0   
2          2001.0           2         608       

In [30]:
# scaling the training dataset first
scaler = MinMaxScaler()
sca_data = scaler.fit_transform(num_data)

# split the data
X_train, X_test, y_train, y_test = train_test_split(
                sca_data, 
                data['SalePrice'],
                test_size = 0.1,
                random_state=0)

In [25]:
# train the model
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.linear_model import Lasso

lin_model = Lasso(alpha=0.001, random_state=0)
# train the model
lin_model.fit(X_train, y_train)

# make predictions for train set
Y_train_pred = lin_model.predict(X_train)

# determine mse, rmse and r2
print('train mse: {}'.format(int(
    mean_squared_error(y_train, Y_train_pred))))
print('train rmse: {}'.format(int(
    mean_squared_error(y_train, Y_train_pred, squared=False))))
print('train r2: {}'.format(
    r2_score(y_train, Y_train_pred)))
print()


# make predictions for test set
Y_test_pred = lin_model.predict(X_test)

# determine mse, rmse and r2
print('test mse: {}'.format(int(
    mean_squared_error(y_test, Y_test_pred))))
print('test rmse: {}'.format(int(
    mean_squared_error(y_test, Y_test_pred, squared=False))))
print('test r2: {}'.format(
    r2_score(y_test, Y_test_pred)))
print()


train mse: 1155694452
train rmse: 33995
train r2: 0.8149062186378196

test mse: 1678003183
test rmse: 40963
test r2: 0.7558243548918225



- train mse: 1183584959
- train rmse: 34403
- train r2: 0.8104393291039478

- test mse: 1672617116
- test rmse: 40897
- test r2: 0.7566081117068992

2. Scaling Later for ML Model

In [39]:
# # split the data
# X_train, X_test, y_train, y_test = train_test_split(
#                 num_data, 
#                 data['SalePrice'],
#                 test_size = 0.1,
#                 random_state=0)

# # scaling the training dataset first
# scaler = MinMaxScaler()
# scaler.fit(X_train)

# X_train = pd.DataFrame(
#     scaler.transform(X_train),
#     columns=X_train.columns
# )

# X_test = pd.DataFrame(
#     scaler.transform(X_test),
#     columns=X_train.columns
# )

In [40]:
# # train the model
# from sklearn.metrics import mean_squared_error, r2_score
# from sklearn.linear_model import Lasso

# lin_model = Lasso(alpha=0.001, random_state=0)
# # train the model
# lin_model.fit(X_train, y_train)

# # make predictions for train set
# pred = lin_model.predict(X_train)

# # determine mse, rmse and r2
# print('train mse: {}'.format(int(
#     mean_squared_error(y_train, pred))))
# print('train rmse: {}'.format(int(
#     mean_squared_error(y_train, pred, squared=False))))
# print('train r2: {}'.format(
#     r2_score(y_train, pred)))
# print()

# # make predictions for test set
# pred = lin_model.predict(X_test)

# # determine mse, rmse and r2
# print('test mse: {}'.format(int(
#     mean_squared_error(y_test, pred))))
# print('test rmse: {}'.format(int(
#     mean_squared_error(y_test, pred, squared=False))))
# print('test r2: {}'.format(
#     r2_score(y_test, pred)))
# print()

train mse: 1155694452
train rmse: 33995
train r2: 0.8149062186378211

test mse: 1678003181
test rmse: 40963
test r2: 0.7558243551769994



- train mse: 1155694452
- train rmse: 33995
- train r2: 0.8149062186378211

- test mse: 1678003181
- test rmse: 40963
- test r2: 0.7558243551769994

3. Scaling First for DL Model

In [31]:
# define all the necessary imports
from keras.models import Sequential
from keras.layers import Dense

In [ ]:
# scaling the training dataset first
scaler = MinMaxScaler()
sca_data = scaler.fit_transform(num_data)

# split the data
X_train, X_test, y_train, y_test = train_test_split(
                sca_data, 
                data['SalePrice'],
                test_size = 0.1,
                random_state=0)

In [33]:
model = Sequential()
# input layer
model.add(Dense(128, kernel_initializer='normal', input_dim= X_train.shape[1], activation='relu' ))

# hidden layers
model.add(Dense(256, kernel_initializer='normal', activation='relu'))
model.add(Dense(256, kernel_initializer='normal', activation='relu'))

# output layer
model.add(Dense(1, kernel_initializer='normal', activation='linear'))

model.compile(loss='mean_absolute_error', optimizer='adam', metrics=['mean_absolute_error'])
model.fit(X_train, y_train, epochs=500, batch_size=32, validation_split = 0.2)

In [36]:
pred = model.predict(X_train)
y_test_pred = model.predict(X_test)

# determine mse, rmse and r2
print('train mse: {}'.format(int(
    mean_squared_error(y_train, pred))))
print('train rmse: {}'.format(int(
    mean_squared_error(y_train, pred, squared=False))))
print('train r2: {}'.format(
    r2_score(y_train, pred)))
print()

# make predictions for test set
pred = lin_model.predict(X_test)

# determine mse, rmse and r2
print('test mse: {}'.format(int(
    mean_squared_error(y_test, pred))))
print('test rmse: {}'.format(int(
    mean_squared_error(y_test, pred, squared=False))))
print('test r2: {}'.format(
    r2_score(y_test, pred)))
print()

train mse: 988943486
train rmse: 31447
train r2: 0.8416127297985978

test mse: 1678003183
test rmse: 40963
test r2: 0.7558243548918225



4. Scaling Laster for DL Model

In [37]:
# split the data
X_train, X_test, y_train, y_test = train_test_split(
                num_data, 
                data['SalePrice'],
                test_size = 0.1,
                random_state=0)

# scaling the training dataset first
scaler = MinMaxScaler()
scaler.fit(X_train)

X_train = pd.DataFrame(
    scaler.transform(X_train),
    columns=X_train.columns
)

X_test = pd.DataFrame(
    scaler.transform(X_test),
    columns=X_train.columns
)

In [38]:
pred = model.predict(X_train)
y_test_pred = model.predict(X_test)

# determine mse, rmse and r2
print('train mse: {}'.format(int(
    mean_squared_error(y_train, pred))))
print('train rmse: {}'.format(int(
    mean_squared_error(y_train, pred, squared=False))))
print('train r2: {}'.format(
    r2_score(y_train, pred)))
print()

# make predictions for test set
pred = lin_model.predict(X_test)

# determine mse, rmse and r2
print('test mse: {}'.format(int(
    mean_squared_error(y_test, pred))))
print('test rmse: {}'.format(int(
    mean_squared_error(y_test, pred, squared=False))))
print('test r2: {}'.format(
    r2_score(y_test, pred)))
print()

train mse: 1009411230
train rmse: 31771
train r2: 0.8383346556498171

test mse: 1651855522
test rmse: 40643
test r2: 0.759629247562508

